### PR TITLE
bpf: switch csum_l4_replace and ipv4_dec_ttl to csum_diff

### PR DESCRIPTION
**Pull request by borkmann**
Moved out of #10491, since it is blocked by #10517.

Another option is to drop it and implement the 2 byte diff in XDP at the expense of more complexity. Need to double check on 4.19.57+ kernels.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/10521)

### CODE REVIEW

**Comment by borkmann** [comment]
test-me-please

File: bpf/lib/csum.h [hunk 1]
```diff
@@ -62,7 +62,13 @@ static __always_inline int csum_l4_replace(struct __ctx_buff *ctx, int l4_off,
 					   struct csum_offset *csum,
 					   __be32 from, __be32 to, int flags)
 {
-	return l4_csum_replace(ctx, l4_off + csum->offset, from, to, flags | csum->flags);
+	__be32 sum = csum_diff(&from, sizeof(from), &to, sizeof(to), 0);
```
Comment (qmonnet) [inline bpf/lib/csum.h]: Nit: You could maybe call `csum_diff_small()` here and avoid to type the two `sizeof()`. Not that it matters much.

**Review by qmonnet — APPROVED** [review]
Looks good!

**Comment by borkmann** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

**Pull request merged**